torch/utils/weak.py

```diff
@@ -1,79 +1,263 @@
+#
+# Our implementation strategy is to create a wrapper weak key object, which we
+# use as a key in a stock Python dictionary.  This is similar to how weakref
+# implements WeakKeyDictionary, but instead of using weakref.ref as the
+# wrapper, we use a custom wrapper that has different __eq__ and __hash__
+# behavior.  Note that we subsequently store this weak key directly in an
+# ORDINARY dictionary, since the newly constructed WeakIdKey's only use would
+# be a dictionary so it would have no strong references.  Ensuring that
+# only live WeakIdKeys are in the map is handled by putting finalizers on the
+# original key object.
 
-# torch.Tensors cannot be used as a key in a dictionary
-# because they define a custom __eq__ function which when used
-# to resolve hash collisions will throw when comparing tensors:
-# "RuntimeError: bool value of Tensor with more than one value is ambiguous."
-# To avoid that, we use an object which will hold a Tensor and use
-# its id for both hashing and equality.
-# In order to use this as a weak key reference, we cannot
-# simply use weakref.WeakKeyDictionary because the newly constructed
-# WeakTensorRefKey only use would be a dictionary so it would have no strong
-# references.
-# To get around this issue, we can use it as a normal key, and then set
-# `weakref.finalize` to delete the key when its contained tensor dies.
 
+# It is simpler to implement this with composition, but if we want to
+# directly reuse the callback mechanism on weakref, we need the weakref
+# and the key to be exactly the same object.  Reusing the callback mechanism
+# minimizes the divergence between our implementation and Lib/weakref.py
+#
+# NB: Prefer using this when working with weakrefs of Tensors; e.g., do
+# WeakIdRef(tensor) rather than weakref.ref(tensor); it handles a number of
+# easy to get wrong cases transparently for you.
+class WeakIdRef(weakref.ref):
+    __slots__ = ['_id']
 
-class WeakTensorRefKey(object):
-    def __init__(self, ten):
-        self.ten = weakref.ref(ten)
-        # store id since as soon as ten is deallocated
-        # the old id will no longer be recoverable, and
-        # we need to be able to remove the WeakTensorRefKey
-        # from the dictionary by hashing it to the same
-        # value it had when ten was alive
-        self.id = id(self.ten())
+    def __init__(self, key, callback=None):
+        # Unlike stock weakref, which preserves hash semantics of the
+        # original object but lazily defers hash calls until the first
+        # time the user attempts to hash the weakref, we can eagerly
+        # cache the id of the key as we know this is definitely the hash
+        # method
+        self._id = id(key)
+        super().__init__(key, callback)
+
+    def __call__(self):
+        r = super().__call__()
+        # Special logic for Tensor PyObject resurrection
+        if hasattr(r, '_fix_weakref'):
+            r._fix_weakref()  # type: ignore[union-attr]
```

Comment:
This class seems to still keep a strong reference to the object. Repro:
```python
import torch
import gc
import weakref
from _torch_weakkeyid import WeakTensorKeyDictionary, WeakIdRef

x = torch.zeros([1])
y = torch.zeros([1])
refx = weakref.ref(x)
refy = weakref.ref(y)
WeakIdRef(x) == WeakIdRef(torch.ones([1]))
del x
del y
gc.collect()
print("refX: ", type(refx())) # refX:  <class 'torch.Tensor'>
print("refY: ", type(refy())) # refY:  <class 'NoneType'>

```

Commenting out "r._fix_weakref()" here fixes it

Reply:
Do you have the first PR in this stack? There was a memory leak that I fixed

Reply:
yeah on master, fixing your example, it works

```
(/Users/ezyang/Dev/pytorch-cpu-env) ezyang-mbp:pytorch-cpu ezyang$ cat foo.py
import torch
import gc
import weakref
from torch.utils.weak import WeakIdRef

x = torch.zeros([1])
y = torch.zeros([1])
refx = weakref.ref(x)
refy = weakref.ref(y)
WeakIdRef(x) == WeakIdRef(torch.ones([1]))
del x
del y
gc.collect()
print("refX: ", type(refx())) # refX:  <class 'torch.Tensor'>
print("refY: ", type(refy())) # refY:  <class 'NoneType'>
(/Users/ezyang/Dev/pytorch-cpu-env) ezyang-mbp:pytorch-cpu ezyang$ python foo.py
refX:  <class 'NoneType'>
refY:  <class 'NoneType'>
```

Reply:
Oh right, thanks a lot! I just imported the `weak.py` file (as I wanted it to work with older versions of pytorch as well). I'll look into the other PR. Thanks!

Reply:
For old versions just get rid of the fix weakref call, it's super busted lol. I suppose if you are willing to write a little C extension code, manually decreffing the Tensor after fixing weakref would work too